pkg/recipes/driver/terraform.go

```diff
@@ -152,7 +152,7 @@ func (d *terraformDriver) Delete(ctx context.Context, opts DeleteOptions) error
 // prepareRecipeResponse populates the recipe response from the module output named "result" and the
 // resources deployed by the Terraform module. The outputs and resources are retrieved from the input Terraform JSON state.
 func (d *terraformDriver) prepareRecipeResponse(ctx context.Context, definition recipes.EnvironmentDefinition, tfState *tfjson.State) (*recipes.RecipeOutput, error) {
-	if tfState == nil || (*tfState == tfjson.State{}) {
+	if tfState == nil || reflect.DeepEqual(*tfState, tfjson.State{}) {
```

Comment:
Was this causing issues before using DeepEqual?

Reply:
Nope. With the updated TF Exec, it started to throw an error.

Reply:
World unit tests be useful here?  To demonstrating that this fixes the issue and explain why a Deep Equal is required.
It's unclear to me at least, what problem this solves and why.

Reply:
+1 to the suggestion from @gpltaylor. Or a comment explaining why this check is important.

Reply:
[This](https://github.com/hashicorp/terraform-json/blob/main/state.go#L41) was added and that is why the equals check started to throw an error.

Reply:
Well this has me worried about all my code, every :) !
So we are guarding against a default implementation. AKA an unpopulated new state object...
Go does not compare Structs (who knew) so DeepEquals can walk the path for us..

My next question is why does the code allow us to get into this bad state branch. However, I'm not qualified to understand how the code gets to this execution point ATM.

So, this is good code IMO 👍 

I made a Gist to get it clear in my head.
https://gist.github.com/gpltaylor/3fe882a8f276f14b15e9a601cd9f6e8f

Reply:
We updated `github.com/hashicorp/terraform-json` from **0.15.0** to **0.19.0** in this PR and we noticed that there was an addition to the **tfjson.State** struct: https://github.com/hashicorp/terraform-json/blob/main/state.go#L41.

Because of this addition our code started to complain about this check: `if tfState == nil || (*tfState == tfjson.State{})`. Now that there is an array in the struct, the code wanted us to do a deeper check, which is the changed line above.

@rynowak both nil and empty check is to make sure that the passed-in state is not nil or empty. It makes sense to me to check for both as we do this type of check in a lot of different places in our code. If we would like more information, maybe @kachawla or @vishwahiremat can add their thoughts here.